File: argus/store.py
import asyncio
import json
import sqlite3
from datetime import UTC, datetime
from enum import StrEnum
from pathlib import Path
from typing import Any
from uuid import uuid4


class RunStatus(StrEnum):
    QUEUED = "queued"
    RUNNING = "running"
    PASSED = "passed"
    FAILED = "failed"
    CANCELLED = "cancelled"


TERMINAL_STATUSES = {RunStatus.PASSED, RunStatus.FAILED, RunStatus.CANCELLED}


def _now() -> str:
    return datetime.now(UTC).isoformat()


class RunStore:
    def __init__(self, database: Path) -> None:
        database.parent.mkdir(parents=True, exist_ok=True)
        self.database = database
        self._lock = asyncio.Lock()

    async def initialize(self) -> None:
        async with self._lock:
            with self._connect() as connection:
                connection.executescript(
                    """
                    CREATE TABLE IF NOT EXISTS runs (
                        id TEXT PRIMARY KEY, url TEXT NOT NULL, instructions TEXT NOT NULL,
                        status TEXT NOT NULL, created_at TEXT NOT NULL, updated_at TEXT NOT NULL,
                        report_json TEXT, error TEXT
                    );
                    CREATE TABLE IF NOT EXISTS events (
                        id INTEGER PRIMARY KEY AUTOINCREMENT, run_id TEXT NOT NULL,
                        type TEXT NOT NULL, data_json TEXT NOT NULL, created_at TEXT NOT NULL,
                        FOREIGN KEY(run_id) REFERENCES runs(id) ON DELETE CASCADE
                    );
                    CREATE INDEX IF NOT EXISTS events_run_id_id ON events(run_id, id);
                    CREATE TABLE IF NOT EXISTS screenshots (
                        id TEXT PRIMARY KEY, run_id TEXT NOT NULL, path TEXT NOT NULL,
                        created_at TEXT NOT NULL,
                        FOREIGN KEY(run_id) REFERENCES runs(id) ON DELETE CASCADE
                    );
                    """
                )

    async def create_run(self, url: str, instructions: str) -> dict[str, Any]:
        run_id, now = uuid4().hex, _now()
        async with self._lock:
            with self._connect() as connection:
                connection.execute(
                    "INSERT INTO runs VALUES (?, ?, ?, ?, ?, ?, NULL, NULL)",
                    (run_id, url, instructions, RunStatus.QUEUED, now, now),
                )
        return await self.get_run(run_id)  # type: ignore[return-value]

    async def list_runs(self, limit: int = 100) -> list[dict[str, Any]]:
        async with self._lock:
            with self._connect() as connection:
                rows = connection.execute(
                    "SELECT * FROM runs ORDER BY created_at DESC LIMIT ?", (limit,)
                ).fetchall()
        return [self._run(row) for row in rows]

    async def get_run(
        self, run_id: str, *, events: bool = False
    ) -> dict[str, Any] | None:
        async with self._lock:
            with self._connect() as connection:
                row = connection.execute(
                    "SELECT * FROM runs WHERE id = ?", (run_id,)
                ).fetchone()
                event_rows = (
                    connection.execute(
                        "SELECT * FROM events WHERE run_id = ? ORDER BY id", (run_id,)
                    ).fetchall()
                    if row is not None and events
                    else []
                )
        if row is None:
            return None
        run = self._run(row)
        if events:
            run["events"] = [self._event(item) for item in event_rows]
        return run

    async def add_event(
        self, run_id: str, event_type: str, data: dict[str, Any] | None = None
    ) -> dict[str, Any]:
        now = _now()
        async with self._lock:
            with self._connect() as connection:
                cursor = connection.execute(
                    "INSERT INTO events(run_id, type, data_json, created_at) VALUES (?, ?, ?, ?)",
                    (run_id, event_type, json.dumps(data or {}), now),
                )
                event_id = cursor.lastrowid
        return {
            "id": event_id,
            "run_id": run_id,
            "type": event_type,
            "data": data or {},
            "created_at": now,
        }

    async def events_after(
        self, run_id: str, event_id: int = 0
    ) -> list[dict[str, Any]]:
        async with self._lock:
            with self._connect() as connection:
                rows = connection.execute(
                    "SELECT * FROM events WHERE run_id = ? AND id > ? ORDER BY id",
                    (run_id, event_id),
                ).fetchall()
        return [self._event(row) for row in rows]

    async def transition(
        self,
        run_id: str,
        *,
        expected: set[RunStatus],
        status: RunStatus,
        event_type: str,
        event_data: dict[str, Any] | None = None,
        report: dict[str, Any] | None = None,
        error: str | None = None,
    ) -> dict[str, Any] | None:
        now = _now()
        expected_values = tuple(expected)
        placeholders = ", ".join("?" for _ in expected_values)
        async with self._lock:
            with self._connect() as connection:
                connection.execute("BEGIN IMMEDIATE")
                cursor = connection.execute(
                    f"""
                    UPDATE runs
                    SET status = ?, updated_at = ?, report_json = COALESCE(?, report_json), error = ?
                    WHERE id = ? AND status IN ({placeholders})
                      AND status NOT IN (?, ?, ?)
                    """,
                    (
                        status,
                        now,
                        json.dumps(report) if report is not None else None,
                        error,
                        run_id,
                        *expected_values,
                        *TERMINAL_STATUSES,
                    ),
                )
                if cursor.rowcount != 1:
                    return None
                event_cursor = connection.execute(
                    "INSERT INTO events(run_id, type, data_json, created_at) VALUES (?, ?, ?, ?)",
                    (run_id, event_type, json.dumps(event_data or {}), now),
                )
        return {
            "id": event_cursor.lastrowid,
            "run_id": run_id,
            "type": event_type,
            "data": event_data or {},
            "created_at": now,
        }

    async def reconcile_interrupted(self) -> list[dict[str, Any]]:
        now = _now()
        data = {"kind": "interrupted", "message": "Run interrupted by server restart"}
        events: list[dict[str, Any]] = []
        async with self._lock:
            with self._connect() as connection:
                connection.execute("BEGIN IMMEDIATE")
                rows = connection.execute(
                    "SELECT id FROM runs WHERE status IN (?, ?)",
                    (RunStatus.QUEUED, RunStatus.RUNNING),
                ).fetchall()
                for row in rows:
                    cursor = connection.execute(
                        """
                        UPDATE runs SET status = ?, updated_at = ?, error = ?
                        WHERE id = ? AND status IN (?, ?)
                        """,
                        (
                            RunStatus.FAILED,
                            now,
                            data["message"],
                            row["id"],
                            RunStatus.QUEUED,
                            RunStatus.RUNNING,
                        ),
                    )
                    if cursor.rowcount != 1:
                        continue
                    event_cursor = connection.execute(
                        "INSERT INTO events(run_id, type, data_json, created_at) VALUES (?, ?, ?, ?)",
                        (row["id"], "run.failed", json.dumps(data), now),
                    )
                    events.append(
                        {
                            "id": event_cursor.lastrowid,
                            "run_id": row["id"],
                            "type": "run.failed",
                            "data": data,
                            "created_at": now,
                        }
                    )
        return events

    async def add_screenshot(self, run_id: str, path: str) -> str:
        screenshot_id = uuid4().hex
        async with self._lock:
            with self._connect() as connection:
                connection.execute(
                    "INSERT INTO screenshots VALUES (?, ?, ?, ?)",
                    (screenshot_id, run_id, path, _now()),
                )
        return screenshot_id

    def _connect(self) -> sqlite3.Connection:
        connection = sqlite3.connect(self.database)
        connection.row_factory = sqlite3.Row
        connection.execute("PRAGMA foreign_keys = ON")
        return connection

    @staticmethod
    def _run(row: sqlite3.Row) -> dict[str, Any]:
        return {
            "id": row["id"],
            "url": row["url"],
            "instructions": row["instructions"],
            "status": row["status"],
            "created_at": row["created_at"],
            "updated_at": row["updated_at"],
            "report": json.loads(row["report_json"]) if row["report_json"] else None,
            "error": row["error"],
        }

    @staticmethod
    def _event(row: sqlite3.Row) -> dict[str, Any]:
        return {
            "id": row["id"],
            "run_id": row["run_id"],
            "type": row["type"],
            "data": json.loads(row["data_json"]),
            "created_at": row["created_at"],
        }

File: argus/internal/store/store.go
package store

import (
	"crypto/rand"
	"database/sql"
	"encoding/hex"
	"encoding/json"
	"os"
	"path/filepath"
	"time"

	"github.com/ace-foundry/argus-testing/argus/internal/domain"
	_ "modernc.org/sqlite"
)

const interruptedMessage = "Run interrupted by server restart"

type Store struct {
	db *sql.DB
}

func Open(path string) (*Store, error) {
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return nil, err
	}
	db, err := sql.Open("sqlite", "file:"+path+"?_pragma=foreign_keys(1)&_pragma=busy_timeout(5000)")
	if err != nil {
		return nil, err
	}
	if err := db.Ping(); err != nil {
		db.Close()
		return nil, err
	}
	return &Store{db: db}, nil
}

func (s *Store) Close() error { return s.db.Close() }

func (s *Store) Initialize() error {
	_, err := s.db.Exec(`
CREATE TABLE IF NOT EXISTS runs (
    id TEXT PRIMARY KEY, url TEXT NOT NULL, instructions TEXT NOT NULL,
    status TEXT NOT NULL, created_at TEXT NOT NULL, updated_at TEXT NOT NULL,
    report_json TEXT, error TEXT
);
CREATE TABLE IF NOT EXISTS events (
    id INTEGER PRIMARY KEY AUTOINCREMENT, run_id TEXT NOT NULL,
    type TEXT NOT NULL, data_json TEXT NOT NULL, created_at TEXT NOT NULL,
    FOREIGN KEY(run_id) REFERENCES runs(id) ON DELETE CASCADE
);
CREATE INDEX IF NOT EXISTS events_run_id_id ON events(run_id, id);
CREATE TABLE IF NOT EXISTS screenshots (
    id TEXT PRIMARY KEY, run_id TEXT NOT NULL, path TEXT NOT NULL,
    created_at TEXT NOT NULL,
    FOREIGN KEY(run_id) REFERENCES runs(id) ON DELETE CASCADE
);`)
	return err
}

func (s *Store) CreateRun(url, instructions string) (*domain.Run, error) {
	id, err := newID()
	if err != nil {
		return nil, err
	}
	now := timestamp()
	if _, err := s.db.Exec(`INSERT INTO runs VALUES (?, ?, ?, ?, ?, ?, NULL, NULL)`, id, url, instructions, domain.RunStatusQueued, now, now); err != nil {
		return nil, err
	}
	return s.GetRun(id, false)
}

func (s *Store) ListRuns(limit int) ([]domain.Run, error) {
	rows, err := s.db.Query(`SELECT * FROM runs ORDER BY created_at DESC LIMIT ?`, limit)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	runs := []domain.Run{}
	for rows.Next() {
		run, err := scanRun(rows)
		if err != nil {
			return nil, err
		}
		runs = append(runs, *run)
	}
	return runs, rows.Err()
}

func (s *Store) GetRun(id string, includeEvents bool) (*domain.Run, error) {
	run, err := scanRun(s.db.QueryRow(`SELECT * FROM runs WHERE id = ?`, id))
	if err == sql.ErrNoRows {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	if includeEvents {
		events, err := s.EventsAfter(id, 0)
		if err != nil {
			return nil, err
		}
		run.Events = events
	}
	return run, nil
}

func (s *Store) AddEvent(runID string, eventType domain.EventType, data map[string]any) (*domain.RunEvent, error) {
	return addEvent(s.db, runID, eventType, data, timestamp())
}

func (s *Store) EventsAfter(runID string, eventID int64) ([]domain.RunEvent, error) {
	rows, err := s.db.Query(`SELECT * FROM events WHERE run_id = ? AND id > ? ORDER BY id`, runID, eventID)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	events := []domain.RunEvent{}
	for rows.Next() {
		event, err := scanEvent(rows)
		if err != nil {
			return nil, err
		}
		events = append(events, *event)
	}
	return events, rows.Err()
}

func (s *Store) Transition(runID string, expected []domain.RunStatus, status domain.RunStatus, eventType domain.EventType, eventData map[string]any, report *domain.RunReport, runError *string) (*domain.RunEvent, error) {
	if len(expected) == 0 || len(expected) > 5 || isTerminal(status) && containsTerminal(expected) {
		return nil, nil
	}
	now := timestamp()
	tx, err := s.db.Begin()
	if err != nil {
		return nil, err
	}
	defer tx.Rollback()
	expectedValues := [5]domain.RunStatus{}
	copy(expectedValues[:], expected)
	args := []any{status, now, reportJSON(report), runError, runID, expectedValues[0], expectedValues[1], expectedValues[2], expectedValues[3], expectedValues[4], domain.RunStatusPassed, domain.RunStatusFailed, domain.RunStatusCancelled}
	result, err := tx.Exec(`UPDATE runs SET status = ?, updated_at = ?, report_json = COALESCE(?, report_json), error = ? WHERE id = ? AND status IN (?, ?, ?, ?, ?) AND status NOT IN (?, ?, ?)`, args...)
	if err != nil {
		return nil, err
	}
	changed, err := result.RowsAffected()
	if err != nil {
		return nil, err
	}
	if changed != 1 {
		return nil, tx.Commit()
	}
	event, err := addEvent(tx, runID, eventType, eventData, now)
	if err != nil {
		return nil, err
	}
	return event, tx.Commit()
}

func (s *Store) ReconcileInterrupted() ([]domain.RunEvent, error) {
	now := timestamp()
	data := map[string]any{"kind": "interrupted", "message": interruptedMessage}
	tx, err := s.db.Begin()
	if err != nil {
		return nil, err
	}
	defer tx.Rollback()
	rows, err := tx.Query(`SELECT id FROM runs WHERE status IN (?, ?)`, domain.RunStatusQueued, domain.RunStatusRunning)
	if err != nil {
		return nil, err
	}
	var ids []string
	for rows.Next() {
		var id string
		if err := rows.Scan(&id); err != nil {
			rows.Close()
			return nil, err
		}
		ids = append(ids, id)
	}
	if err := rows.Close(); err != nil {
		return nil, err
	}
	events := make([]domain.RunEvent, 0, len(ids))
	for _, id := range ids {
		result, err := tx.Exec(`UPDATE runs SET status = ?, updated_at = ?, error = ? WHERE id = ? AND status IN (?, ?)`, domain.RunStatusFailed, now, interruptedMessage, id, domain.RunStatusQueued, domain.RunStatusRunning)
		if err != nil {
			return nil, err
		}
		changed, err := result.RowsAffected()
		if err != nil {
			return nil, err
		}
		if changed != 1 {
			continue
		}
		event, err := addEvent(tx, id, domain.EventRunFailed, data, now)
		if err != nil {
			return nil, err
		}
		events = append(events, *event)
	}
	return events, tx.Commit()
}

func (s *Store) AddScreenshot(runID, path string) (string, error) {
	id, err := newID()
	if err != nil {
		return "", err
	}
	_, err = s.db.Exec(`INSERT INTO screenshots VALUES (?, ?, ?, ?)`, id, runID, path, timestamp())
	return id, err
}

type rowScanner interface{ Scan(...any) error }
type eventExecutor interface {
	Exec(string, ...any) (sql.Result, error)
}

func scanRun(row rowScanner) (*domain.Run, error) {
	var run domain.Run
	var report sql.NullString
	var runError sql.NullString
	if err := row.Scan(&run.ID, &run.URL, &run.Instructions, &run.Status, &run.CreatedAt, &run.UpdatedAt, &report, &runError); err != nil {
		return nil, err
	}
	if report.Valid {
		var value domain.RunReport
		if err := json.Unmarshal([]byte(report.String), &value); err != nil {
			return nil, err
		}
		run.Report = &value
	}
	if runError.Valid {
		value := runError.String
		run.Error = &value
	}
	return &run, nil
}

func scanEvent(row rowScanner) (*domain.RunEvent, error) {
	var event domain.RunEvent
	var data string
	if err := row.Scan(&event.ID, &event.RunID, &event.Type, &data, &event.CreatedAt); err != nil {
		return nil, err
	}
	if err := json.Unmarshal([]byte(data), &event.Data); err != nil {
		return nil, err
	}
	return &event, nil
}

func addEvent(executor eventExecutor, runID string, eventType domain.EventType, data map[string]any, now string) (*domain.RunEvent, error) {
	if data == nil {
		data = map[string]any{}
	}
	encoded, err := json.Marshal(data)
	if err != nil {
		return nil, err
	}
	result, err := executor.Exec(`INSERT INTO events(run_id, type, data_json, created_at) VALUES (?, ?, ?, ?)`, runID, eventType, encoded, now)
	if err != nil {
		return nil, err
	}
	id, err := result.LastInsertId()
	if err != nil {
		return nil, err
	}
	return &domain.RunEvent{ID: id, RunID: runID, Type: eventType, Data: data, CreatedAt: now}, nil
}

func reportJSON(report *domain.RunReport) any {
	if report == nil {
		return nil
	}
	encoded, _ := json.Marshal(report)
	return string(encoded)
}

func isTerminal(status domain.RunStatus) bool {
	return status == domain.RunStatusPassed || status == domain.RunStatusFailed || status == domain.RunStatusCancelled
}
func containsTerminal(statuses []domain.RunStatus) bool {
	for _, status := range statuses {
		if isTerminal(status) {
			return true
		}
	}
	return false
}
func timestamp() string { return time.Now().UTC().Format("2006-01-02T15:04:05.000000+00:00") }
func newID() (string, error) {
	value := make([]byte, 16)
	if _, err := rand.Read(value); err != nil {
		return "", err
	}
	return hex.EncodeToString(value), nil
}
